### ISSUE TITLE
Determine GPUExternalTexture lifetime

### ISSUE THREAD

**Issue by Kangz** [gpuweb/gpuweb#2124]
It is important that `importExternalTexture` works with the ubiquitous `HTMLVideoElement` as it does today in the spec. However for browsers that implement WebCodec we should also be able to import `VideoFrame` as `GPUExternalTexture`. Since the API will take a `VideoFrame`, it will simply not possible to use it when WebCodec isn't present since you wouldn't be able to create a `VideoFrame`.

I think the only change in the API needed is the following:

```diff
dictionary GPUExternalTextureDescriptor : GPUObjectDescriptorBase {
-   required HTMLVideoElement source;
+   required (HTMLVideoElement | VideoFrame) source;
    // ...
};
```

Then we also need to specify what's the lifetime of the resulting `GPUExternalTexture`. Is it destroyed on `VideoFrame.close()`?

FYI @shaoboyan 

[EDIT by @kainino0x]: Investigation: #1380

**Comment by shaoboyan** [comment]
@Kangz  Thanks for proposing this. `VideoFrame.close()` is at least one of the signal to destroy `GPUExternalTexture`. And current lifecycle logic could still work for `GPUExternalTexture`.

**Comment by kainino0x** [comment]
I'm finally thinking about this. Here are two possible reasonable options for the lifetimes, considering an example 3d video application that has e.g. a 24Hz video and a 60Hz render rate.

1. No GPUExternalTexture.close().
    - When created from VideoFrame, closing the VideoFrame closes the GPUExternalTexture. It doesn't close earlier.
    - When created from `<video>`, it gets closed automatically in a microtask as today. App would simply re-import every time they want to use the video, but browsers likely want to cache one frame to avoid unnecessary re-importing work.
    - Applications that need more control than provided by importing `<video>` must use VideoFrame.

2. Add GPUExternalTexture.close() but keep the auto-closing semantics when importing from `<video>`. (Close would just close early.)
    - When created from VideoFrame, closing either the VideoFrame or the GPUExternalTexture does NOT close the other. Both must be closed explicitly.
    - Still requires caching as above.

3. Add GPUExternalTexture.close() and change the semantics of imports from `<video>`.
    - When created from `<video>`, it must be closed explicitly (causing a warning if it is GCed before closed). App would have to implement caching itself (if it wants it), by watching for `requestVideoFrameCallback` callbacks (or possibly by watching the currentTime attribute for changes, but IIRC this is imprecise?) and closing the previous frame when a new one is available.
    - This is more explicit and somewhat matches the style of the rest of WebGPU better.
    - Applications could stall a `<video>` by holding onto too many imported frames from the decoder pool. (It is no worse than [Chromium's implementation of] `new VideoFrame(video_element)`.) If the browser uses a decoder ring buffer instead of a decoder pool (does any?), holding just one frame for too long could stall the decoder.
    - Requires browser to be able to hold strong-refs to video frames. Easy for Chromium (we do this in VideoFrame), but not sure about other browsers.

Needless to say I'm leaning away from option 3, despite the browser caching in 1/2. But maybe there is more to consider here.
Among 1/2, I slightly prefer 1, as early-closing should be unnecessary and it avoids adding a `.close()` method.

**Comment by shaoboyan** [comment]
I'm a fan of option 1.  
For option 2, I feel a bit weird about "closing either the VideoFrame or the GPUExternalTexture does NOT close the other." 
Option3 is too complicated and may stall the decoder.

For option 1, from the impl view, we need to have a mechanism to ensure VideoFrame.close() will notify WebGPU. So we could  report errors for below scenario:
```
const externalTextureDescriptor = { source: videoFrame};
const externalTexture = device.importExternalTexture(externalTextureDescriptor);
...
videoFrame.close();
...
 const bindGroup = device.createBindGroup({
    layout: pipeline.getBindGroupLayout(0),
    entries: [
      {
        binding: 0,
        resource: externalTexture, // should report validation error
      },
    ],
  });
```

**Comment by Kangz** [comment]
It seems there's two orthogonal questions that may not need to be resolved the same way.

A) How is the lifetime of imported `<video>` managed? We have three solutions outlined so far:

 1. The application explicitly calls `GPUExternalTexture.destroy/close()` and risks stalling the video decoder if it doesn't do it correctly.
 2. The `GPUExternalTexture` is automatically closed at the end of the microtask. Requires caching on the browser side.
 3. The `GPUExternalTexture` is closed when the callback passed to `requestVideoFrameCallback` is called. No caching is required, but WebGPU takes a dependency on browsers implementing `requestVideoFrameCallback`.

B) How is the lifetime of imported `VideoFrame` managed? Solutions so far:

 1. There is only `VideoFrame.close()` that also controls the lifetime of the `GPUExternalTexture`.
 2. Both `GPUExternalTexture.destroy/close()`  and `VideoFrame.close()` close the `GPUExternalTexture`. 
 3. `VideoFrame.close()` doesn't do anything on the `GPUExternalTexture`, only `GPUExternalTexture.destroy/close()` does.

I think A3 would be ideal but it depends on what other browsers feel about `requestVideoFrame`. Either B1 or B2 seems ok, I have concerns with B3 because it means that either WebCodec or WebGPU can stall the video decoder.

**Comment by kainino0x** [comment]
For A3 I suppose we could just expose a GPUExternalTexture."is still alive" attribute (guaranteed not to change inside a task). This would expose duplicate information with rVFC but is much more straightforward/obvious and doesn't require rVFC support. My concern either way is that it's rather fragile, and pages could end up depending on lifetimes that don't turn out to be portable. Maybe it works if the spec is constraining enough, like, say it can only change on rAF boundaries just after the rAF callback? I don't know if there are any constraints today on when rVFC can fire.

So for this reason I think I still like A2. No real opinion about B; as long as we put in the right warnings, developers will adapt easily to any of these.

**Comment by jdashg** [comment]
I like A2.
I think the pre-standards aspect of WebCodec makes B hard to sign up for/agree to.

**Comment by Kangz** [comment]
> My concern either way is that it's rather fragile, and pages could end up depending on lifetimes that don't turn out to be portable.

This seems like a minor concern in the face of usability. There are already 1000 ways promises / callbacks get scheduled differently on the Web platform between browsers, and developers almost always get things correct. Over-specifying the scheduling of things takes some control away from browsers (like tab freezing / throttling, etc).

> I like A2.

Please describe why you like it better than others, otherwise it's hard to discuss alternatives.

> I think the pre-standards aspect of WebCodec makes B hard to sign up for/agree to.

https://github.com/mozilla/standards-positions/issues/209 describes this as "worth prototyping" and Mozilla is very involved in WebCodec (and even has a group chair). In all cases I think we should design this interaction with the assumption that it can be changed at any time until WebCodec is a stable specification. It's not possible to access the functionality without WebCodec being available anyway so there is no breaking change in changing it when WebCodec isn't a stable shipped spec.

And this is somewhat similar to WebXR / WebGPU integration where we already did design work (though it happened in the immersive web group) for the integration between an unreleased API, and an API that's not in all browsers.

**Comment by kainino0x** [comment]
> This seems like a minor concern in the face of usability.

I have mixed thoughts on how much A2 presents usability issues. It will be a bit counterintuitive to users (and potentially perform poorly if there isn't good caching), but the behavior is very reliable which I think will quickly correct any developer misunderstandings.

I was mostly concerned about timing hazards, but upon further discussion with @Kangz I think we can eliminate that issue with stricter timing:

> Maybe it works if the spec is constraining enough, like, say it can only change on rAF boundaries just after the rAF callback? I don't know if there are any constraints today on when rVFC can fire.

I still don't know how exactly when rVFC can fire, but I don't think it matters. At least for Chromium, it's definitely possible to hold strong references to video frames (as WebCodecs VideoFrame does this). So simply: the lifetime lasts from importExternalTexture until the end of some future rAF (in particular, the one after the next rVFC).

The user does still need to know when to re-import. rVFC actually might be too often(?) depending on its behavior when e.g. playing a 60Hz video on a 30Hz display. But regardless, it would be handy to have an alternate signal like "is still alive": this cuts the dependency on rVFC and is more practical for rendering use-cases anyway (`if (not still alive) texture = device.importExternalTexture(video);`).

One minor concern: this sets no expectation of caching in the browser, yet applications could still import too often. I think we could guard against this with a warning if the same video frame is imported multiple times (perhaps in different animation frames). Developers would see such a warning reliably since video framerate is almost always lower than display framerate.

**Comment by shaoboyan** [comment]
> The GPUExternalTexture is closed when the callback passed to requestVideoFrameCallback is called. No caching is required, but WebGPU takes a dependency on browsers implementing requestVideoFrameCallback.

If I understand correctly, I think this means we could only create (or suggest to creating) `GPUExternalTexture` in `requestVideoFrameCallback` and the benefit is that browser don't need to cache the frame.

I think "bullet screen" is one of the cases that graphics needs to working in display fps but not the video fps. When we watching "bullet screen", we could always find some fast bullets which move across the screen smoothly. I think rendering the animation of the bullets with display fps will provide better experience(means, the rendering logic should work in rAF). So if we requires user to create GPUExternalTexture in `rVFC` and user needs to render bullets in `rAF`, the logic is a bit complex. It might be:
```
const bulletTexture = device.CreateTexture(...);

window.requestAnimationFrame(function renderingBullets() {
 //rendering bulletTexture in another draw
});

video.requestVideoFrameCallback(function updateExternalTexture() {
   const externalTexture = device.importExternalTexture(...)
  // rendering the video frame firstly.
});
```

Nowadays, "bullet screen" also have a popular tech calls "anti-mask"(the translation maybe not correct) which the bullets won't cover the human face or important objects in video frame. Instead it flies behind them. For me, it is something like video frame itself has different depth from different parts. Then app composites them with bullets. I'm not sure above code logic could handle this effectively (Here is [a showcase](https://www.bilibili.com/video/BV15W411o755/) if you have interests).


And having a cache in browser seems a good solution to improve the performance. In Chromium, WebGL also cache one frame to improve `texImage2D(HTMLVideoElement)` performance.  The problem is that is it possible to achieve 0-copy with the cache ? I think it might be possible because we could keep the single frame from being updated until new frame has been decoded or decoder tells WebGPU that they need to release the frame.

> The GPUExternalTexture is automatically closed at the end of the microtask. Requires caching on the browser side.

I prefer A2 for importing `<video>`. Constraint the lifecycle of `GPUExternalTexture` into the microtask is reasonable. If user wants to cache the frame, they could do it by rendering the frame into a texture. Destory() function seems not useful.

> There is only VideoFrame.close() that also controls the lifetime of the GPUExternalTexture.

I prefer B1 for importing `VideoFrame`. For me, If `VideoFrame.Close()` doesn't bother `GPUExternalTexture`, it means that the `GPUExternalTexture` is a copy of "VideoFrame".  That is not what I expected. I think `GPUExternalTexture` is a external object stub in WebGPU. If the external object has been destroyed, the stub should be invalid too.

**Comment by Kangz** [comment]
Let me rephrase the proposal @kainino0x made above, and let's name it **A4**:

`GPUExternalTextures` from `HTMLVideoElement` can only be invalidated at the end of rAF. When first imported, `GPUExternalTextures` are guaranteed to be valid at least until the end of the next rAF (or the current one if we're in a rAF callback). `GPUExternalTexture` expose some member that says whether they are valid. Like `.valid`.

This means that developers can use the following code:

```js
var videoTexture = null;
var videoBindGroup = null;
function frame() {
    if (!videoTexture || !videoTexture.valid) {
        videoTexture = device.importExternalTexture({source: myVideo});
        videoBindGroup = device.createBindGroup(...);
    }

    // Do something with the video bind group

    requestAnimationFrame(frame);
}

requestAnimationFrame(frame);
```

> I prefer A2 for importing <video>. Constraint the lifecycle of GPUExternalTexture into the microtask is reasonable. If user wants to cache the frame, they could do it by rendering the frame into a texture. Destory() function seems not useful.

While this is a workable solution, I think it makes things more difficult for developers if they want to use external texture in more asynchronous code. It's also better to design a solution that doesn't require a cache in the browser if possible because it reduces implementation complexity and makes things simpler to reason about for developers.

**Comment by kainino0x** [comment]
> If user wants to cache the frame, they could do it by rendering the frame into a texture.

This would be unnecessarily costly because it would always create an extra copy. If the browser performs the caching it would have the option to the video frame alive and cache a simple wrapper instead.

I suppose this is an example of the confusing nature of A2. It's hard to tell whether app-side copy or repeated re-importing is better.

**Comment by shaoboyan** [comment]
> Let me rephrase the proposal @kainino0x made above, and let's name it A4:
Ok, my point is that I don't like to constraint `GPUExternalTexture` into `rVFC` which is different from other rendering logics. I think `A4` is a balance for browser doesn't need extra cache and user doesn't to cache the thing themseleves, and seems great.

**Comment by kainino0x** [comment]
Unfortunately the reality of videos is that they're highly asynchronous and don't schedule nicely with rAF. I think we're in agreement that A3 (in which the import ends at rVFC) is not ideal.

I'll open a PR soon to propose A4 (after a little thinking about how to expose the "is valid" flag).

Also, that should give us some clarity on VideoFrame since it doesn't add a destroy method, so regardless of whether we spec it here or not I think it's reasonable to settle on option B1.

**Comment by kainino0x** [comment]
@grorg @jdashg Can you provide feedback on whether proposal A4 will work for you, specifically:
Can you hold an HTMLVideoElement video frame (which could be a decoder pool resource) alive for a browser-controlled amount of time that would be no longer than one rAF after the video element has advanced past that frame?

**Comment by kainino0x** [comment]
FTR the corresponding investigation for VideoFrame is #1380.

**Comment by litherum** [comment]
Can this issue be renamed to something that doesn't use the word "WebCodec"? During the last call it was stated that it doesn't actually have anything to do with WebCodec.

**Comment by kainino0x** [comment]
There are two issues here, one is WebCodecs and one isn't.

**Comment by kainino0x** [comment]
The lifetime issue came up again here because it was blocking the VideoFrame API design.

**Comment by kainino0x** [comment]
> I don't know if there are any constraints today on when rVFC can fire.

There are. As drafted (https://wicg.github.io/video-rvfc/#video-rvfc-procedures):

> Note: The effective rate at which callbacks are run is the lesser rate between the video’s rate and the browser’s rate. When the video rate is lower than the browser rate, the callbacks' rate is limited by the frequency at which new frames are presented. When the video rate is greater than the browser rate, the callbacks' rate is limited by the frequency of the update the rendering steps. This means, a 25fps video playing in a browser that paints at 60Hz would fire callbacks at 25Hz; a 120fps video in that same 60Hz browser would fire callbacks at 60Hz.

**Comment by litherum** [comment]
Our video decoders decode into a pool of IOSurfaces. An IOSurface will be re-used from the pool if it has a reference count of 1 (meaning the pool is the only thing referencing it). If there is no such IOSurface in the pool, a new one is allocated.

IOSurfaces are "wired" kernel memory, which means the kernel's VM system can't relocate the addresses or pages. Therefore, they require more maintenance care than regular allocations.

This means that we would be willing to allocate _a few_ extra IOSurfaces if JS retains them, but we would not be willing to give JS carte blanche to retain as many of them as it feels like.

**Comment by kainino0x** [comment]
Thanks @litherum! That shouldn't be a problem for the proposal I'm working on in #2302.

**Comment by kainino0x** [comment]
Drafted w3c/webcodecs#412 to consider adding the spec to WebCodecs.

**Comment by kainino0x** [comment]
Resolved to open that draft. I'll take care of driving it on the webcodecs side.

With that plus #2302 this issue will be closed!

**Comment by dalecurtis** [comment]
FWIW, the WebCodec's spec allows implementations to stall decoding for any reason. The most common being that the client has exhausted the allowed supply of hardware decoding buffers. So WebKit would be free to limit the number of outstanding IOSurfaces to whatever it feels is appropriate.

**Comment by kainino0x** [comment]
Feedback on https://github.com/w3c/webcodecs/pull/412 has been negative on including this functionality in WebCodecs when it is clearly a part of WebGPU that interacts with WebCodecs and not the other way around. From that thread:

> It's fine not to implement a specification in its entirety, it's fairly common.

Though I agree this is common, I know Apple has taken a hard stance against this as they intend to implement 100% of the WebGPU spec with no omissions. @litherum how do you view the tradeoff between that intent and the problem of "monkey patching" WebGPU from the WebCodecs spec?

**Comment by Kangz** [comment]
WebCodec should be a different issue.

**Comment by kainino0x** [comment]
Filed #2498 for WebCodecs. Closing this since this issue has too many mixed topics. #2302 is still open but we can track it on that PR instead of here.
